File: 2024/elixir/day1.exs
input = File.read!("../inputs/day1.txt")

list1 = []
list2 = []

lines = String.split(input, "\n")

{list1,list2} = Enum.reduce(lines,{list1,list2}, fn line,{list1,list2} ->
  nums = String.split(line, "   ")
  if nums != [""] do
    [first | rest] = nums
    {list1 ++ [first], list2 ++ rest}
  else
    {list1,list2}
  end
end)

list1 = Enum.map(list1,fn i -> String.to_integer(i) end)
list2 = Enum.map(list2,fn i -> String.to_integer(i) end)

list1 = Enum.sort(list1)
list2 = Enum.sort(list2)

ans = 0

ans = Enum.reduce(Enum.zip(list1,list2),ans,fn list,ans ->
  {a,b} = list
  ans + abs(a-b)
end)

IO.puts("The answer for part 1 is: #{ans}")

map = %{}

map = Enum.reduce(list2,map,fn i,map ->
    if map[i] == nil do
      map = Map.put(map,i,1)
      map
    else
      map = Map.put(map,i,map[i]+1)
      map
    end
end)

ans = 0

ans = Enum.reduce(list1,ans,fn i,ans ->
  if map[i]==nil do
    ans
  else
    ans + i * map[i]
  end
end)

IO.puts("The answer for part 2 is: #{ans}")

File: 2024/rust/src/bin/day1.rs
use::std::fs;
use::std::collections::HashMap;

fn main() -> Result<(), Box<dyn std::error::Error>>{

    let input = fs::read_to_string("../inputs/day1.txt")?;

    let mut list1: Vec<i64> = Vec::new();
    let mut list2: Vec<i64> = Vec::new();

    for line in input.lines(){
        let nums:Vec<&str> = line.split_whitespace().collect();
        list1.push(nums[0].parse()?);
        list2.push(nums[1].parse()?);
    }

    list1.sort();
    list2.sort();

    let mut ans:i64 = 0;

    for i in 0..list1.len(){
        ans+=(list1[i]-list2[i]).abs();
    }

    println!("The answer for part 1 is: {}", ans);

    let mut counter:HashMap<i64, i64> = HashMap::new();

    for i in list2.iter(){
        *counter.entry(*i).or_insert(0)+=1;
    }

    let mut ans:i64 = 0;

    for i in list1.iter(){
        if counter.contains_key(i){
            ans+=(*i)*counter[i];
        }
    }

    println!("The answer for part 2 is: {}", ans);

    Ok(())
}
